net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,316 @@
+ *	          
+ *          https://github.com/AltraMayor/gatekeeper
+ *
+ *
+ */
+
+
+struct gkprioq_sched_data {
+/* Parameters */
+	u32 max_limit;
+
+/* Variables */
+	struct Qdisc *prio_array[MAX_PRIORITY];
+	u16 highest_prio;
+};
+
+static inline u16 calc_new_high_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	for (prio = q->highest_prio - 1; prio > 0; prio--) {
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+
+	return 0; /* Queue 0 could be non empty OR GK queue is empty. In any
+		   * case return 0 (default highest priority setting). */
+}
+
+static int gkprioq_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct Qdisc *your_queue;
+	int prio, ret, wlen, gk_len;
+	u16 your_prio; 
+
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);
+
+	switch(tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv4_get_dsfield(ip_hdr(skb))) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+	 	wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv6_get_dsfield(ipv6_hdr(skb))) >> 2;
+	      	break;
+
+	default: 
+		your_prio = 0;
+		break;
+	}
+
+	if (your_prio > q->highest_prio)
+		q->highest_prio = your_prio; /* Checking if a new highest 
+					      * priority. */
+			
+	your_queue = q->prio_array[your_prio];
+
+	gk_len = sch->q.qlen;
+	if (gk_len < q->max_limit) {
+
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+		return ret;
+		}
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* your_queue is taking up all of GK, no choice but to remove a packet
+	 * and then enqueue the ingressed packet.
+	 */
+
+	else if (your_queue->q.qlen == your_queue->limit || your_prio == 0) {
+
+		__qdisc_queue_drop_head(your_queue, &your_queue->q, to_free);
+		qdisc_qstats_drop(sch);
+
+		ret = qdisc_enqueue(skb, your_queue, to_free);
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+				return ret;
+		}		
+
+		qdisc_qstats_backlog_inc(sch, skb);
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* Look to lower priority queues to check if ingress packet can be 
+	 * accomodated. If no such queues exist, no choice but to drop
+	 * ingress packet.
+	 */
+
+	else {
+		prio = 0;			
+		while (prio < your_prio) { 
+			if (q->prio_array[prio]->q.qlen > 0) { 		
+		
+				__qdisc_queue_drop_head(q->prio_array[prio], 
+					&(q->prio_array[prio])->q, to_free);
+				qdisc_qstats_drop(sch);
+
+
+				ret = qdisc_enqueue(skb, your_queue, to_free);
+				if (ret != NET_XMIT_SUCCESS) {
+					if (net_xmit_drop_count(ret))
+						qdisc_qstats_drop(sch);
+						return ret;
+				}
+
+				qdisc_qstats_backlog_inc(sch, skb);
+				sch->q.qlen++;
```

Comment:
We shouldn't, `__qdisc_drop_head` doesn't update the stats on its own, so yes, I've removed the `sch->q.qlen++`  now, thanks.